crates/core/src/operations/put.rs

```diff
@@ -838,148 +860,62 @@ pub enum PutState {
-                }
-            } else {
-                // New contract, use PutQuery which will handle it properly
-                tracing::debug!("Contract {} is new, storing locally", key);
-                put_contract(
-                    op_manager,
-                    key,
-                    value.clone(),
-                    related_contracts.clone(),
-                    contract,
-                )
-                .await?;
-            }
-
-            // Mark operation as successful
-            put_op.state = Some(PutState::Finished { key });
-            return Ok(());
-        } else {
-            return Err(OpError::UnexpectedOpState);
-        }
-    }
-
-    let target = target.unwrap();
-
-    tracing::debug!("PUT: Selected target {} for contract {}", target.peer, key);
-
-    let id = put_op.id;
-    if let Some(stats) = &mut put_op.stats {
-        stats.target = Some(target.clone());
-    }
-
-    match put_op.state {
+    // Process PrepareRequest state and transition to next state
+    let (id, contract, value, related_contracts, htl, subscribe, target) = match &put_op.state {
         Some(PutState::PrepareRequest {
             contract,
-            related_contracts,
             value,
+            related_contracts,
             htl,
             subscribe,
         }) => {
-            let new_state = Some(PutState::AwaitingResponse {
-                key,
-                upstream: None,
-                contract: contract.clone(),
-                state: value.clone(),
-                subscribe,
-            });
-            let msg = PutMsg::RequestPut {
-                id,
-                contract,
-                related_contracts,
-                value,
-                htl,
-                target: target.clone(),
-            };
-
-            let op = PutOp {
-                id,
-                state: new_state,
-                stats: put_op.stats,
-            };
-
-            op_manager
-                .notify_op_change(NetMessage::from(msg), OpEnum::Put(op))
-                .await?;
+            let key = contract.key();
+            let own_location = op_manager.ring.connection_manager.own_location();
+
+            // Find the optimal target for this contract
+            let target = op_manager
+                .ring
+                .closest_potentially_caching(&key, [&own_location.peer].as_slice())
+                .unwrap_or(own_location); // If no other peers, target ourselves
+
+            (
+                put_op.id,
+                contract.clone(),
+                value.clone(),
+                related_contracts.clone(),
+                *htl,
+                *subscribe,
+                target,
+            )
         }
-        _ => return Err(OpError::invalid_transition(put_op.id)),
-    }
+        _ => return Err(OpError::UnexpectedOpState),
+    };
+
+    // Transition to AwaitingResponse state (similar to GET operation)
```

Comment:
what happens with the logic that used to check if the contract existed and updated it in that case? I don't think we are handling that correctly in the AwaitingResponse state so we are losing that logic here, is that ok?

Reply:
this hasn't been addressed